bitcoin/shadouble.c

```diff
@@ -1,18 +1,23 @@
-#include "shadouble.h"
+#include <bitcoin/shadouble.h>
 #include <bitcoin/groestl.h>
 #include <ccan/mem/mem.h>
 #include <common/type_to_string.h>
 #include <common/utils.h>
 
+void groestl512_double(struct sha256_double *shadouble, const void *p, size_t len)
+{
+	groestlhash((void *)&shadouble->sha, (void *)p, len);
+}
+
 void sha256_double(struct sha256_double *shadouble, const void *p, size_t len)
 {
 	sha256(&shadouble->sha, memcheck(p, len), len);
-	groestlhash((void *)&shadouble->sha, (void *)p, len);
+	sha256(&shadouble->sha, &shadouble->sha, sizeof(shadouble->sha));
 }
 
 void sha256_double_done(struct sha256_ctx *shactx, struct sha256_double *res)
 {
-	/* jush finish to generate single sha for groestl */
+	/* jush finish to generate single sha for groestl tx*/
```

Comment:
@dflatemaster @gruve-p I think `sha256_double_done` should be reverted to upstream version as well:
1. It is very confusing and error prone that `sha256_double_done` computes single SHA despite name. Especially when `sha256_double` actually computes double SHA as advertised.
2. Hash is replaced everywhere, not only in places which are required for Groestlcoin Core compatibility.

It suggest to introduce `sha256_done` and update code in required places to use it.

Reply:
https://github.com/Groestlcoin/lightning/pull/16#issuecomment-514405608

Reply:
@yura-pakhuchiy What do you mean with `sha256_double_done` should be reverted to upstream? Upstream uses `sha256_double_done` also https://github.com/ElementsProject/lightning/blob/master/bitcoin/shadouble.c#L11

1. It is confusing indeed however I think @dflatemaster did it this way to make it easy for us to merge with upstream each release (otherwise you would have more merge conflicts)

2. Hash is only replaced here https://github.com/Groestlcoin/lightning/blob/master/bitcoin/tx.c#L270

If we really want to change it then I would propose to change `sha256_double_done` to `single_sha256_done` . Feel free to send a PR with the suggested fixes.

Reply:
@gruve-p I mean that upstream version of `sha256_double_done` makes double SHA, but our makes single SHA and by reverting I mean we should should do double SHA in this function.
@dflatemaster's version maybe will be produce less merge conflicts, but it will be highly fragile. What point of merge without conflicts when resulting code does not work?

Reply:
https://github.com/Groestlcoin/lightning/tree/single_sha256_done something like this?

Reply:
Yes, this looks a lot better.

Reply:
Alternative version on https://github.com/Groestlcoin/lightning/pull/17 Please review